// @yll10243/vue3-carousel/dist/index.js
import { defineComponent as X, reactive as C, onMounted as Y, onBeforeUnmount as E, watch as _, openBlock as c, createElementBlock as d, createElementVNode as B, Fragment as w, renderList as I, normalizeClass as S, renderSlot as L, createCommentVNode as M } from "vue";
const N = { class: "d-carousel-inner" }, F = {
  key: 0,
  class: "d-carousel-indicators"
}, V = ["onClick"], $ = /* @__PURE__ */ X({
  name: "AyiCarousel",
  __name: "index",
  props: {
    /**
     * 轮播的数据源
     */
    listData: {
      type: Array,
      default() {
        return [];
      }
    },
    /**
     * 是否显示地步的指示器，默认是true
     */
    showIndicator: {
      type: Boolean,
      default: !0
    },
    /**
     * 是否自动滚动，默认是true
     */
    autoPlay: {
      type: Boolean,
      default: !0
    },
    /**
     * 滚动到下一张的时间间隔
     */
    autoPlayTime: {
      type: Number,
      default: 2e3
    },
    /**
     * 进入的动画
     */
    inAnimate: {
      type: String,
      default: "fadeIn"
    },
    inAnimateOtherSide: {
      type: String,
      default: "fadeIn"
    },
    /**
     * 出去的动画
     */
    outAnimate: {
      type: String,
      default: "fadeOut"
    },
    outAnimateOtherSide: {
      type: String,
      default: "fadeOut"
    },
    /**
     * 速度
     */
    speed: {
      type: String,
      default: ""
    },
    /**
     * 等待时间
     */
    waitingTime: {
      type: Number,
      default: 500
    }
  },
  setup(i, { expose: g }) {
    const t = i, p = t.listData.map(() => "hide");
    p[0] = "";
    const a = C({
      activeIndex: 0,
      isAutoPlay: t.autoPlay,
      playing: !1,
      itemClasses: p
    }), v = [], f = C({ pageX: 0, pageY: 0 }), A = () => {
      for (let e of v)
        clearTimeout(e);
    }, m = () => {
      a.isAutoPlay = !1, A();
    }, D = () => {
      if (A(), !a.isAutoPlay || a.playing)
        return;
      let e = setTimeout(() => {
        a.isAutoPlay && y();
      }, t.autoPlayTime);
      v.push(e);
    }, r = () => {
      a.isAutoPlay = t.autoPlay, D();
    }, T = (e) => {
      e < 0 ? e = t.listData.length - 1 : e >= t.listData.length && (e = 0), a.activeIndex = e;
    }, P = () => {
      !a.playing && T(a.activeIndex - 1);
    }, y = () => {
      !a.playing && T(a.activeIndex + 1);
    }, O = (e, n) => {
      if (!t.listData.length || a.playing) return;
      m(), a.playing = !0;
      const s = t.listData.length - 1;
      let o = e < n || e === s && n === 0;
      e === 0 && n === s && (o = !1);
      const u = [];
      for (let l = 0; l < t.listData.length; l++)
        l === e ? u[l] = [o ? t.inAnimateOtherSide : t.inAnimate, "active"] : l === n ? u[l] = [o ? t.outAnimateOtherSide : t.outAnimate] : u[l] = ["hide"], u[l].push(t.speed);
      a.itemClasses = u, t.waitingTime ? setTimeout(() => {
        a.playing = !1, r();
      }, t.waitingTime) : (a.playing = !1, r());
    }, k = (e) => {
      m();
      const { pageX: n, pageY: s } = e.changedTouches[0];
      f.pageX = n, f.pageY = s;
    }, b = (e) => {
      const n = f.pageX - e.changedTouches[0].pageX, s = f.pageY - e.changedTouches[0].pageY;
      (Math.abs(s) > Math.abs(n) ? s : n) > 0 ? y() : P();
    };
    return Y(() => {
      r(), window.addEventListener("blur", m), window.addEventListener("focus", r);
    }), E(() => {
      window.removeEventListener("focus", r), window.removeEventListener("blur", m);
    }), _(
      [() => t.listData, () => t.autoPlay],
      () => {
        r();
      },
      {
        deep: !0,
        //对层次复制的对象
        immediate: !0
        //是否立即执行
      }
    ), _(
      () => a.activeIndex,
      (e, n) => {
        O(e, n);
      },
      {
        deep: !0,
        //对层次复制的对象
        immediate: !0
        //是否立即执行
      }
    ), g({
      prev: P,
      next: y
    }), (e, n) => (c(), d("div", {
      class: "d-carousel",
      onMouseenter: m,
      onMouseleave: r,
      onTouchstart: k,
      onTouchend: b
    }, [
      B("div", N, [
        (c(!0), d(w, null, I(i.listData, (s, o) => (c(), d("div", {
          key: o,
          class: S(["item animated", a.itemClasses[o]])
        }, [
          L(e.$slots, "default", {
            record: s,
            index: o
          }, void 0, !0)
        ], 2))), 128))
      ]),
      i.showIndicator ? (c(), d("div", F, [
        (c(!0), d(w, null, I(i.listData.length, (s, o) => (c(), d("div", {
          class: S(["item", { active: o === a.activeIndex }]),
          key: o,
          onClick: (u) => a.activeIndex = o
        }, null, 10, V))), 128))
      ])) : M("", !0)
    ], 32));
  }
}), x = (i, g) => {
  const t = i.__vccOpts || i;
  for (const [p, a] of g)
    t[p] = a;
  return t;
}, h = /* @__PURE__ */ x($, [["__scopeId", "data-v-49f13bcc"]]), R = {
  install(i) {
    i.component(h.name ? h.name : "AyiCarousel", h);
  }
};
export {
  R as default
};
